### PR TITLE
Add biometrics support to browser_sign_in sample

### PR DESCRIPTION
**Pull request by rajdeepnanua-okta**

### CODE REVIEW

File: browser-sign-in/src/main/java/com/okta/android/samples/browser_sign_in/storage/SerializableToken.kt [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        tokenType = tokenType,
+        expiresIn = expiresIn,
+        accessToken = accessToken,
+        scope = scope,
+        refreshToken = refreshToken,
+        idToken = idToken,
+        deviceSecret = deviceSecret,
+        issuedTokenType = issuedTokenType,
+    )
+}
```
[the remark this answers is not in the document]
Comment (mikenachbaur-okta) [reply browser-sign-in/src/main/java/com/okta/android/samples/browser_sign_in/storage/SerializableToken.kt:59]: Why did you need to copy the code here?

**Review by mikenachbaur-okta — APPROVED** [review]
I have a couple small questions, but it's nothing big. I like what you've done here.

What I would like to see is some code comments in the relevant places that explains to the developer what you're doing, and why.  I wouldn't want a customer to think they have to implement all of this, all the time.

File: browser-sign-in/src/main/java/com/okta/android/samples/browser_sign_in/storage/SerializableToken.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (rajdeepnanua-okta) [reply browser-sign-in/src/main/java/com/okta/android/samples/browser_sign_in/storage/SerializableToken.kt:59]: The code I copied is internal API. I just ended up using the same serialization library as okta-mobile-kotlin (kotlinx-serialization), so it was easier to just recycle the SDK's internal code for serialization purposes.

**Comment by rajeshjadidminc** [comment]
java.security.KeyStoreException: the master key android-keystore://com_okta_sample_storage exists but is unusable
                                                                                                    	at com.google.crypto.tink.integration.android.AndroidKeysetManager$Builder.readOrGenerateNewMasterKey(AndroidKeysetManager.java:276)
                                                                                                    	at com.google.crypto.tink.integration.android.AndroidKeysetManager$Builder.build(AndroidKeysetManager.java:237)
                                                                                                    	at androidx.security.crypto.EncryptedSharedPreferences.create(EncryptedSharedPreferences.java:165)
                                                                                                    	at com.cfna.app.okta.storage.SharedPreferencesModule.createSharedPreferences(SharedPreferencesModule.kt:47)
                                                                                                    	at com.cfna.app.okta.storage.SharedPreferencesModule.providesBiometricCredentialSharedPrefs(SharedPreferencesModule.kt:94)
                                                                                                    	at com.cfna.app.okta.storage.SharedPreferencesModule_ProvidesBiometricCredentialSharedPrefsFactory.providesBiometricCredentialSharedPrefs(SharedPreferencesModule_ProvidesBiometricCredentialSharedPrefsFactory.java:42)
                                                                                                    	at com.cfna.app.DaggerCFNAApplication_HiltComponents_SingletonC$SingletonCImpl$SwitchingProvider.get(DaggerCFNAApplication_HiltComponents_SingletonC.java:757)
                                                                                                    	at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
                                                                                                    	at com.cfna.app.okta.biometric.BiometricCredentialsManager.useBiometricCredentialStorage(BiometricCredentialsManager.kt:64)
                                                                                                    	at com.cfna.app.ui.home.HomeKt$launchBiometricPrompt$2$onAuthenticationSucceeded$1.invokeSuspend(Home.kt:269)
                                                                                                    	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
                                                                                                    	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
                                                                                                    	at androidx.compose.ui.platform.AndroidUiDispatcher.performTrampolineDispatch(AndroidUiDispatcher.android.kt:81)
                                                                                                    	at androidx.compose.ui.platform.AndroidUiDispatcher.access$performTrampolineDispatch(AndroidUiDispatcher.android.kt:41)
                                                                                                    	at androidx.compose.ui.platform.AndroidUiDispatcher$dispatchCallback$1.run(AndroidUiDispatcher.android.kt:57)
                                                                                                    	at android.os.Handler.handleCallback(Handler.java:938)
                                                                                                    	at android.os.Handler.dispatchMessage(Handler.java:99)
                                                                                                    	at android.os.Looper.loop(Looper.java:233)
                                                                                                    	at android.app.ActivityThread.main(ActivityThread.java:8068)
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method)
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:631)
                                                                                                    	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:978)
                                                                                                    	Suppressed: kotlinx.coroutines.DiagnosticCoroutineContextException: [androidx.compose.ui.platform.MotionDurationScaleImpl@f6488b6, androidx.compose.runtime.BroadcastFrameClock@239d9b7, StandaloneCoroutine{Cancelling}@43057b4, AndroidUiDispatcher@e5c788d]
                                                                                                    Caused by: android.security.keystore.UserNotAuthenticatedException: User not authenticated
                                                                                                    	at android.security.KeyStore.getInvalidKeyException(KeyStore.java:1369)
                                                                                                    	at android.security.KeyStore.getInvalidKeyException(KeyStore.java:1388)
                                                                                                    	at android.security.keystore.KeyStoreCryptoOperationUtils.getInvalidKeyExceptionForInit(KeyStoreCryptoOperationUtils.java:54)
                                                                                                    	at android.security.keystore.KeyStoreCryptoOperationUtils.getExceptionForCipherInit(KeyStoreCryptoOperationUtils.java:89)
                                                                                                    	at android.security.keystore.AndroidKeyStoreCipherSpiBase.ensureKeystoreOperationInitialized(AndroidKeyStoreCipherSpiBase.java:265)
                                                                                                    	at android.security.keystore.AndroidKeyStoreCipherSpiBase.engineInit(AndroidKeyStoreCipherSpiBase.java:109)
                                                                                                    	at javax.crypto.Cipher.tryTransformWithProvider(Cipher.java:2984)
                                                                                                    	at javax.crypto.Cipher.tryCombinations(Cipher.java:2891)
                                                                                                    	at javax.crypto.Cipher$SpiAndProviderUpdater.updateAndGetSpiAndProvider(Cipher.java:2796)
                                                                                                    	at javax.crypto.Cipher.chooseProvider(Cipher.java:773)
                                                                                                    	at javax.crypto.Cipher.init(Cipher.java:1143)
                                                                                                    	at javax.crypto.Cipher.init(Cipher.java:1084)
                                                                                                    	at com.google.crypto.tink.integration.android.AndroidKeystoreAesGcm.encryptInternal(AndroidKeystoreAesGcm.java:84)
                                                                                                    	at com.google.crypto.tink.integration.android.AndroidKeystoreAesGcm.encrypt(AndroidKeystoreAesGcm.java:72)
2022-12-23 21:19:03.914 15506-15506 AndroidRuntime          com.cfna.app.debug                   E  	at com.google.crypto.tink.integration.android.AndroidKeystoreKmsClient.validateAead(AndroidKeystoreKmsClient.java:259)
                                                                                                    	at com.google.crypto.tink.integration.android.AndroidKeystoreKmsClient.getAead(AndroidKeystoreKmsClient.java:175)
                                                                                                    	at com.google.crypto.tink.integration.android.AndroidKeysetManager$Builder.readOrGenerateNewMasterKey(AndroidKeysetManager.java:268)
                                                                                                    	... 21 more

### Reproduction Steps?

- Login without biometric (remove biometric setup from device)
- Go dashboard
- Enable checkbox( i have managed the code with if biometric is not enable then it will go setting and come back with enabling the biometric and check again in activity result if its enable run the same which you have mention for checkbox)
- application crashed.

File: browser-sign-in/src/main/java/com/okta/android/samples/browser_sign_in/storage/SerializableToken.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (mikenachbaur-okta) [reply browser-sign-in/src/main/java/com/okta/android/samples/browser_sign_in/storage/SerializableToken.kt:59]: Lets make a reminder to clean this up, when we update token storage later.

**Comment by rajdeepnanua-okta** [comment]
@rajeshjadidminc, I believe I have addressed that issue with the latest commit to this PR. Let me know if you're still running into this issue after that, and let's discuss the issue further in https://github.com/okta/samples-android/issues/99

**Comment by rajeshjadidminc** [comment]
@rajdeepnanua-okta

I have take the master branch pull and tril to run the code, i m getting below error:


com.okta.authfoundation.client.OidcClientResult$Error$OidcEndpointsNotAvailableException: OIDC Endpoints not available.
                                                                                                    	at com.okta.authfoundation.client.OidcClient.endpointNotAvailableError(OidcClient.kt:257)
                                                                                                    	at com.okta.oauth2.AuthorizationCodeFlow.start$oauth2_release(AuthorizationCodeFlow.kt:129)
                                                                                                    	at com.okta.oauth2.AuthorizationCodeFlow$start$2.invokeSuspend(Unknown Source:20)
                                                                                                    	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
                                                                                                    	at kotlinx.coroutines.internal.ScopeCoroutine.afterResume(Scopes.kt:33)
                                                                                                    	at kotlinx.coroutines.AbstractCoroutine.resumeWith(AbstractCoroutine.kt:102)
                                                                                                    	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
                                                                                                    	at kotlinx.coroutines.DispatchedTaskKt.resume(DispatchedTask.kt:234)
                                                                                                    	at kotlinx.coroutines.DispatchedTaskKt.resumeUnconfined(DispatchedTask.kt:190)
                                                                                                    	at kotlinx.coroutines.DispatchedTaskKt.dispatch(DispatchedTask.kt:161)
                                                                                                    	at kotlinx.coroutines.CancellableContinuationImpl.dispatchResume(CancellableContinuationImpl.kt:397)
                                                                                                    	at kotlinx.coroutines.CancellableContinuationImpl.resumeImpl(CancellableContinuationImpl.kt:431)
                                                                                                    	at kotlinx.coroutines.CancellableContinuationImpl.resumeImpl$default(CancellableContinuationImpl.kt:420)
                                                                                                    	at kotlinx.coroutines.CancellableContinuationImpl.resumeWith(CancellableContinuationImpl.kt:328)
                                                                                                    	at kotlinx.coroutines.ResumeAwaitOnCompletion.invoke(JobSupport.kt:1413)
                                                                                                    	at kotlinx.coroutines.JobSupport.notifyCompletion(JobSupport.kt:1520)
                                                                                                    	at kotlinx.coroutines.JobSupport.completeStateFinalization(JobSupport.kt:323)
                                                                                                    	at kotlinx.coroutines.JobSupport.finalizeFinishingState(JobSupport.kt:240)
                                                                                                    	at kotlinx.coroutines.JobSupport.tryMakeCompletingSlowPath(JobSupport.kt:906)
                                                                                                    	at kotlinx.coroutines.JobSupport.tryMakeCompleting(JobSupport.kt:863)
                                                                                                    	at kotlinx.coroutines.JobSupport.makeCompletingOnce$kotlinx_coroutines_core(JobSupport.kt:828)
                                                                                                    	at kotlinx.coroutines.AbstractCoroutine.resumeWith(AbstractCoroutine.kt:100)
                                                                                                    	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
                                                                                                    	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
                                                                                                    	at android.os.Handler.handleCallback(Handler.java:938)
                                                                                                    	at android.os.Handler.dispatchMessage(Handler.java:99)
                                                                                                    	at android.os.Looper.loop(Looper.java:233)
                                                                                                    	at android.app.ActivityThread.main(ActivityThread.java:8068)
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method)
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:631)
                                                                                                    	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:978)

**Comment by rajeshjadidminc** [comment]
Hi,

I have applied same changes as per the master branch but getting below crash:

FATAL EXCEPTION: main
                                                                                                    Process: com.cfna.app.debug, PID: 22511
                                                                                                    java.lang.IllegalStateException: Attempted accessing credential storage before calling setSharedPreferences
                                                                                                    	at com.cfna.app.okta.storage.CredentialTokenStorage$accessStorage$2.invokeSuspend(CredentialTokenStorage.kt:89)
                                                                                                    	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
                                                                                                    	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
                                                                                                    	at kotlinx.coroutines.internal.LimitedDispatcher.run(LimitedDispatcher.kt:42)
                                                                                                    	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:95)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:570)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:677)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:664)
                                                                                                    	Suppressed: kotlinx.coroutines.DiagnosticCoroutineContextException: [StandaloneCoroutine{Cancelling}@adf88c6, Dispatchers.Main.immediate]

**Pull request merged**